src/base/ClpMap.h

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (C) 1996-2020 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID__SRC_BASE_CLPMAP_H
+#define SQUID__SRC_BASE_CLPMAP_H
+
+#include "SquidTime.h"
+
+#include <list>
+#include <map>
+
+template<class EntryValue>
+size_t
+DefaultMemoryUsage(const EntryValue *e)
+{
+    return sizeof(*e);
+}
+
+/**
+ * A Map for caching data by Capacity, Lifetime, and Priority (CLP)
+ * Unlike other Map containers data is;
+ * - added only if it fits within a predetermined memory limit (Capacity),
+ * - gets expired based on TTL (Lifetime), and a fading Priority Queue.
+ */
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *) = DefaultMemoryUsage>
+class ClpMap
+{
+public:
+    class Entry
+    {
+    public:
+        Entry(const Key &aKey, EntryValue *t): key(aKey), value(t), date(squid_curtime) {}
+        ~Entry() {delete value;}
+    private:
+        Entry(Entry &);
+        Entry & operator = (Entry &);
+    public:
+        Key key; ///< the key of entry
+        EntryValue *value = nullptr; ///< A pointer to the stored value
+        time_t date = 0; ///< The date the entry created
+    };
+    typedef std::list<Entry *> Queue;
```

Comment:
To avoid unnecessary raw memory allocations, this type needs adjustments. I have not done the research necessary to find the best type. The usual suspects include:

1. `std::list<Entry>` with an allocator
1. `std::list<Entry*>` with an allocator (with a justification for doubling the number of allocations/dereferences: `STL list item wrapper->entry pointer->stuffWeActuallyNeed`)
1. an invasive list of pointers to Entry

Please also document the order this Queue is maintaining.

Reply:
I have done some research. It lends towards using dqueue for the final storage. My primary goal is getting this CLP mapping available for use within Squid. If you want to add the storage optimization as part of this PR though I am open to that.

FTR:

* The required operations are: push to head, keyed search, pop random, pop tail - in that order of commonality.

* For the operations most commonly performed on caches the choice we have is between list, dequeue, vector, map. Other containers I'm aware of lack the ability to perform some required operation.

* From the (circa 2012) speed comparisons I was able to find std::dqueue outperforms list on almost all fronts, and vector on the primary push to head case. No data on std::map there - I imagine that since moves are done via pop+push the lookup hash management is a killer cost.

  * push to head - list and dqueue are equal with essentially zero time.

  * keyed search - vector or dqueue pretty close.

  * remove random - list is fastest by a huge margin which grows as object sizes increase. second best is dqueue. vector gets very, very bad.

  * remove tail - list and dqueue and vector are equal with essentially zero time.

Reply:
> It lends towards using dqueue for the final storage

What exactly do you mean by `dqueue`? std::deque?

> My primary goal is getting this CLP mapping available for use within Squid. If you want to add the storage optimization as part of this PR though I am open to that.

* If you are adding a new class, it should have adequate performance for the intended use cases. My usual [principles](https://github.com/measurement-factory/squid-notes/blob/start/review.md) of requiring PR cleanup would apply to every class aspect in this case, including storage.
* If you are enhancing LruMap with new functionality, then the PR acceptance criteria would be different because the PR would only be responsible for new/changed code.

Based on the PR description and the diff, it feels like you are doing the former, but I am ready to re-evaluate this PR if your intent is the latter. Please clarify this PR scope/intent.

I will come back to this change request once `dqueue` definition and PR scope are clear.

Reply:
This is functionality upgrade to LruMap.

Reply:
> This is functionality upgrade to LruMap.

I interpret "functionality upgrade to LruMap" as enhancing LruMap by adding new LruMap functionality and leaving the old LruMap code/functionality unchanged to the extent possible. This PR is not, by default, responsible for improving existing LruMap code/storage/etc. -- it reuses what it can safely reuse. The unusable PR diff is just an unfortunate side effect of premature renaming. Please correct me if my interpretation is wrong.

> It lends towards using dqueue for the final storage.

What exactly do you mean by dqueue? std::deque?

Also, please clarify what you mean be "it" or "it lends" in this context. My change request identified unnecessary raw memory allocations. You responded with the above phrase. I do not see a way to interpret "lends towards using dqueue for the final storage" as justifying those unnecessary raw memory allocations, but perhaps you were commenting on something else?

Reply:
WTF, my reply to you went in as an edit rather than a quote. sorry about that.

Reply:
>> It lends towards using dqueue for the final storage.
>
> What exactly do you mean by dqueue? std::deque?

Yes.

> Also, please clarify what you mean be "it" or "it lends" in this context.

The benchmarks I found when researching the STL container performances. That showed list as very sub-par in the most common operations performed by our Lru/ClpMap template.

 > My change request identified unnecessary raw memory allocations. You responded with the above phrase. I do not see a way to interpret "lends towards using dqueue for the final storage" as justifying those unnecessary raw memory allocations, but perhaps you were commenting on something else?

If we are going to widely change the storage related code of this template to optimize the existing internal storage. Then we should move to the container with optimal performance at the same time. Which is not the std::list currently used.

Reply:
It feels like your evaluation in earlier [comment](https://github.com/squid-cache/squid/pull/593/files#r404478061) was not for the type we are discussing here -- Queue. This (currently undocumented) type is for ensuring the right LRU removal order. Queue does not participate in item _searches_. We have Map for that. For Queue, the operations are:

* push_back() -- adding a new item to the cache;
* pop_front() -- purging the next capacity victim from the cache;
* erase(iterator) -- removing a caller-requested item from the cache.

std::deque, std::list, and a custom invasive list (all with an allocator) have similar (amortised) small costs for the first two operations. std::deque loses a bit on the erase() speed, but that is not why std::deque has to be disqualified. The problem with std::deque is that its erase() _invalidates all iterators_ (that are stored in Map). We cannot afford that kind of damage because we have to keep Queue and Map in sync!

This leaves us with std::list with an allocator (item 1 on my [earlier](https://github.com/squid-cache/squid/pull/593/files#r404089699) list) and a custom invasive list (item 3 on my earlier list). I recommend the former because it is an existing, well-known class with overheads similar to a custom one (AFAICT). Please adjust the PR code accordingly.

Reply:
Allocator added in commit 52b7adb8abbe686e2bb6856f267b01c7842e7187

Reply:
Moved to https://github.com/squid-cache/squid/pull/593#discussion_r420145859